File: core/src/com/desitum/library/view/Button.java
package com.desitum.library.view;

import com.badlogic.gdx.graphics.g2d.TextureRegion;
import com.desitum.library.drawing.Drawable;
import com.desitum.library.game.World;

/**
 * Created by kodyvanry on 5/1/17.
 */

public class Button extends View {

    // DRAWING
    private Drawable mRestTexture;
    private Drawable mHoverTexture;

    public Button(World world) {
        this(world, null);
    }

    public Button(World world, LayoutConstraints layoutConstraints) {
        super(world, layoutConstraints);
        setClickable(true);
    }

    @Override
    public boolean onTouchEvent(TouchEvent touchEvent) {
        switch (touchEvent.getAction()) {
            case DOWN:
                if (mHoverTexture != null)
                    setBackgroundDrawable(mHoverTexture);
                break;
            case MOVE:
                if (isTouching(touchEvent) && mHoverTexture != null)
                    setBackgroundDrawable(mHoverTexture);
                else if (mRestTexture != null)
                    setBackgroundDrawable(mRestTexture);
                break;
            case UP:
                if (mRestTexture != null)
                    setBackgroundDrawable(mRestTexture);
        }
        return super.onTouchEvent(touchEvent);
    }

    public void setRestDrawable(Drawable restDrawable) {
        this.mRestTexture = restDrawable;
        if (getBackgroundDrawable() == null)
            setBackgroundDrawable(this.mRestTexture);
    }

    public void setHoverDrawable(Drawable hoverDrawable) {
        this.mHoverTexture = hoverDrawable;
        if (getBackgroundDrawable() == null)
            setBackgroundDrawable(this.mHoverTexture);
    }
}

File: core/src/com/desitum/library/view/Button.kt
package com.desitum.library.view

import com.desitum.library.drawing.Drawable
import com.desitum.library.game.AssetManager
import com.desitum.library.game.World

/**
 * Created by kodyvanry on 5/1/17.
 */

open class Button @JvmOverloads constructor(world: World, layoutConstraints: LayoutConstraints? = null) : View(world, layoutConstraints) {

    // DRAWING
    var restDrawable: Drawable? = null
        set(value) {
            backgroundDrawable = backgroundDrawable ?: value
            field = value
        }

    private var _restDrawable: Int = -1


    var hoverDrawable: Drawable? = null
        set(value) {
            backgroundDrawable = backgroundDrawable ?: value
            field = value
        }

    init {
        clickable = true
        if (_restDrawable != -1)
            restDrawable = AssetManager.instance.getDrawable(_restDrawable)
    }

    override fun onTouchEvent(touchEvent: TouchEvent): Boolean {
        when (touchEvent.action) {
            TouchEvent.Action.DOWN -> if (hoverDrawable != null)
                backgroundDrawable = hoverDrawable
            TouchEvent.Action.MOVE -> if (isTouching(touchEvent) && hoverDrawable != null)
                backgroundDrawable = hoverDrawable
            else -> backgroundDrawable = restDrawable
        }
        return super.onTouchEvent(touchEvent)
    }
}
